include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,712 @@
+ *  std::vector<double>  data_a        = {2, 3, 5, 6};
+ *
+ *  std::vector<hsize_t> coordinates_b = {5,
+ *                                        0, // first point
+ *                                        0,
+ *                                        4, // second point
+ *                                        5,
+ *                                        4, // third point
+ *                                        26,
+ *                                        12}; // fourth point
+ *  std::vector<double>  data_b        = {9, 4, 7, 6};
+ *  if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+ *    {
+ *      dataset.write_selection(data_a, coordinates_a);
+ *    }
+ *  else if (Utilities::MPI::this_mpi_process(mpi_communicator) == 1)
+ *    {
+ *      dataset.write_selection(data_b, coordinates_b);
+ *    }
+ *  else
+ *    {
+ *      dataset.write_none<double>();
+ *    }
+ * @endcode
+ *
+ * ## Query the type of I/O that HDF5 performed on the last parallel I/O call
+ * In some cases such as when there is type conversion the HDF5 library can
+ * decide to do independent I/O instead of collective I/O. The following code
+ * can be used to query the I/O method.
+ * @code
+ * auto dataset = group.create_dataset<double>("name", dimensions);
+ * #ifdef DEBUG
+ * dataset.check_io_mode(true);
+ * #endif
+ *
+ * if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+ *     {
+ *       dataset.write(data);
+ *     }
+ * else
+ *     {
+ *       dataset.write_none<double>();
+ *     }
+ *
+ * #ifdef DEBUG
+ * pcout << "IO mode :" << dataset.io_mode<std::string>() << std::endl;
+ * pcout << "Local no collective cause :"
+ *       << dataset.local_no_collective_cause<std::string>() << std::endl;
+ * pcout << "Global no collective cause :"
+ *       << dataset.global_no_collective_cause<std::string>() << std::endl;
+ * #endif
+ * @endcode
+ *
+ * @author Daniel Garcia-Sanchez, 2018
+ */
+namespace HDF5
+
+{
+  /**
+   * Base class for the HDF5 objects.
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class HDF5Object
+  {
+  protected:
+    HDF5Object(const std::string name, bool mpi);
+
+  public:
+    enum class Mode
+    {
+      create,
+      open
+    };
+
+    /**
+     * Reads an attribute. T can be float, double, std::complex<float>,
+     * std::complex<double>, int, unsigned int, bool or std::string. Note that
+     * the encoding of std::string is UTF8 in order to be compatible with
+     * python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    T
+    attr(const std::string attr_name);
```

Comment:
I used `attr` in order to be consistent with the API of [python](http://docs.h5py.org/en/stable/high/attr.html), but I can use the deal.ii conventions :)

Reply:
Yes, let's stick with the deal.II convention. The point is that you want to write the class in such a way that it can be used in user code without knowing all that much about HDF5 to begin with, and in particular without having to learn the HDF5 naming conventions -- after all, that's sort of the point of wrappers: To hide a library behind something that looks and feels like deal.II ;-)